python_ta/typecheck/base.py

```diff
@@ -129,42 +133,83 @@ def lookup_method(name, caller_type, *args):
     return TYPE_SIGNATURES[caller_origin][name]
 
 
+def make_set(node):
+    node.parent = node
+    # node.rank = 0
+
+
+def _find_rep(node):
+    while node.parent != node:
+        node = node.parent
+    return node
+
+
+def _union(node1, node2):
+    rep1 = _find_rep(node1)
+    rep2 = _find_rep(node2)
+    rep2.parent = rep1
+
+
+class TNode:
+    def __init__(self, node_type, origin_node):
+        self.type = node_type
+        self.origin = origin_node
+
+    def __str__(self):
+        return self._type.__name__
+
+
 class TypeConstraints:
     """Represents all the type constraints in the program."""
     def __init__(self):
         self._count = 0
         self._sets = []
+        self._tvar_tnode = {}
 
     def clear_tvars(self):
         """Resets the type constraints kept track of in the program."""
         self._count = 0
         self._sets = []
+        self._tvar_tnode = {}
 
-
-    def fresh_tvar(self) -> TypeVar:
-        """Return a fresh type variable."""
+    def fresh_tvar(self, node) -> TypeVar:
+        """Return a fresh type variable with the node it was created in."""
         tvar = TypeVar('_T' + str(self._count))
-        self._sets.append({tvar})
+        tnode = TNode(tvar, node)
+        self._sets.append(tnode)
+        make_set(tnode)
+        self._tvar_tnode[tvar] = tnode
         self._count += 1
         return tvar
 
-    def _find(self, t: TypeVar) -> int:
-        """Return the index of the set containing t."""
-        for i, type_set in enumerate(self._sets):
-            if t in type_set:
-                return i
-        return -1
+    def add_concrete_to_sets(self, _type):
```

Comment:
Unify is not be guaranteed to take in TypeVar instances as arguments as you mentioned above, so I need to be able to create TNodes representing the concrete types in the sets and have a way to look them up. I agree that these nodes should not be stored in our _tvars_tnode dictionary so I decided to create a new private dictionary attribute in TypeConstraints to store the concrete types and their corresponding TNodes. This requires a helper because I do not want to raise the counts attribute while creating these concrete type Nodes and I also do not want to add them to the tvars dictionary.